CIP-DappConnector/CIP-dapp-connector.md

```diff
@@ -0,0 +1,269 @@
+
+### api.signData(addr: cbor\<address>, sigStructure: cbor\<Sig_structure>): Promise\<Bytes>
+
+Errors: `APIError`, `DataSignError`
+
+This endpoint utilizes the [CIP-0008 signing spec](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0008/CIP-0008.md) for standardization/safety reasons. It allows the dApp to request the user to sign data conforming to said spec. The user's consent should be requested and the details of `sig_structure` shown to them in an informative way. The  Please refer to the CIP-0008 spec for details on how to construct the sig structure.
+
+### api.submitTx(tx: cbor\<transaction>): Promise\<hash32>
+
+Errors: `APIError`, `TxSendError`
+
+As wallets should already have this ability, we allow dApps to request that a transaction be sent through it. If the wallet accepts the transaction and tries to send it, it shall return the transaction id for the dApp to track. The wallet is free to return the `TxSendError` with code `Refused` if they do not wish to send it, or `Failure` if there was an error in sending it (e.g. preliminary checks failed on signatures).
+
+## Events
+
+In addition to the API methods that are actively called, the connector also must emit the following events. All methods events are required to be implemented.
+
+TODO: event emission method? Possible methods:
+
+1) Emitted event via `window.addEventListener(eventName , e => void)`
+2) Emitted message via `window.postMessage({eventName: string}, ...)`
+3) Some kind of callback registration i.e. `wallet.onDisconnect(() => void)` or `wallet.onEvent(eventName => void)`
+4) A combination of the two (event/message but with callback on `wallet` object as well
+5) Other?
+
+### wallet_disconnected
+
+Emitted when the user disconnects (not just changes) their current account from the page. This means that all `api` methods will return with an `APIError.Refused` error and a new `api` object must be obtained from `wallet.enable()` to continue to interact with the user's wallet.
+
+### account_changed
+
+Emitted when the user changes accounts (i.e. different root key pair and/or network id). The same `api` object will continue to work but will now return results based on the new account. After this event is received dApps should check `api.getNetworkId()` as changing accounts can also change the network.
+
+
```

Comment:
I would add another event for network_changed in case that the user switches from mainnet to testnet in the wallet

Reply:
It was discussed in other comments above, but it went something like, why should this be separate from account changed? If you have the same recovery phrase but your wallet lets you switch between testnets without the "account" changing in the wallet side, it's still effectively changing since all the addresses, UTXOs, balance, and basically everything is going to change too. The dApp probably shouldn't care that the key phrase remains the same but different network (thus different state entirely) vs changing between two unrelated key phrases on different networks, unless you can provide a use-case that justifies it.